src/HieDb/Utils.hs

```diff
@@ -38,25 +38,39 @@ import Data.IORef
 import HieDb.Types
 import HieDb.Compat
 import Database.SQLite.Simple
+import Control.Monad.State.Strict (StateT, get, put)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Control.Monad (when)
 
 #if __GLASGOW_HASKELL__ >= 903
 import Control.Concurrent.MVar (readMVar)
 #endif
 
-addTypeRef :: HieDb -> FilePath -> A.Array TypeIndex HieTypeFlat -> A.Array TypeIndex (Maybe Int64) -> RealSrcSpan -> TypeIndex -> IO ()
+-- Each AST Node can have a lot of repetitive type information.
+-- We use this to make sure that each occurrence of a type
+-- (Identified by Int64 id of that type within typenames table)
+-- is inserted exactly once per each RealSrcSpan in which it occurs and per
+-- depth of that type within the tree structure representing the type (Int)
+type TypeIndexing a = StateT (Set (RealSrcSpan, Int64, Int)) IO a
```

Comment:
I'm pretty sure we can do better than this structure.
Given how the traversal is progressing (e.g. RealSrcSpan is really constant throughout this function) it might be faster to use something like 

`Map RealSrcSpan (Set (Int6, Int))`

Reply:
we only need StateT around `mapM_ (addTypeRef db path arr ixs (nodeSpan ast))` because each node in the AST will have a different span.

Reply:
Will try this change. But is this safe? Is every AST node guaranteed to have different span?

Reply:
yes, the way the AST is constructed, all nodes with the same span are merged together.

Reply:
Ok, changed in https://github.com/wz1000/HieDb/pull/67/commits/c4562a1b7d9adb35064f53c895342f811840f552